src/grpc/stream_infer_handler.cc

```diff
@@ -610,11 +611,12 @@ ModelStreamInferHandler::StreamInferResponseComplete(
     // If this was the final callback for the state
     // then cycle through the completion queue so
     // that state object can be released.
-    if (state->complete_) {
+    if (is_complete) {
       state->step_ = Steps::CANCELLED;
       state->context_->PutTaskBackToQueue(state);
     }
 
+    state->complete_ = is_complete;
```

Comment:
Based on the root cause - Is the use of local variable sufficient? Or is there some synchronization needed?

Reply:
I think this is sufficient, because it is only a bool variable and only moves from `false` to `true` (unless the function is called with a complete final flag and then called again without the flag, which should not happen). Given the call to this function with the complete final flag is the last call into this function, it is safe to mark the state as complete at the end, so that the gRPC thread can capture the complete signal and release the state.